ion/symbols_table_test.go

```diff
@@ -0,0 +1,268 @@
+	}
+
+	return symbolTableCase{
+		desc:    desc,
+		kind:    kind,
+		name:    name,
+		version: version,
+		imports: imports,
+		maxSID:  sid,
+		tokens:  tokens,
+		table:   *table,
+	}
+}
+
+func makeSharedSymbolTableCase(
+	desc string, name string, version int32, texts []string, imports ...SymbolTable) symbolTableCase {
+	makeToken := func(text string, sid int64) SymbolToken {
+		source := ImportSource{
+			Table: name,
+			SID:   sid,
+		}
+		return SymbolToken{
+			Text:     newString(text),
+			localSID: sid,
+			Source:   &source,
+		}
+	}
+	newTable := func(imports ...SymbolTable) *SymbolTable {
+		return newSharedSymbolTable(name, version, imports...)
+	}
+	return makeSymbolTableCase(
+		desc, 0, symbolTableKindShared, name, version, makeToken, newTable, texts, imports...)
+}
+
+func makeLocalSymbolTableCase(desc string, texts []string, imports ...SymbolTable) symbolTableCase {
+	makeToken := func(text string, sid int64) SymbolToken {
+		return SymbolToken{
+			Text:     newString(text),
+			localSID: sid,
+			Source:   nil,
+		}
+	}
+	return makeSymbolTableCase(
+		desc, 9, symbolTableKindLocal, "", 0, makeToken, newLocalSymbolTable, texts, imports...)
+}
+
+// Arbitrary, deterministic seed for our symbol table testing
+const testSystemSymbolTableSeed = 0x03CB815D3119751B
+
+func TestSymbolTable(t *testing.T) {
+	// TODO test tables with imports
+	cases := []symbolTableCase{
+		makeSystemSymbolTableCase(),
+		makeLocalSymbolTableCase("Empty LST", nil),
+		makeLocalSymbolTableCase("Simple LST", []string{"a", "b", "c"}),
+		makeSharedSymbolTableCase("Empty SST", "foo", 1, nil),
+		makeSharedSymbolTableCase("Simple SST", "bar", 3, []string{"cat", "dog", "moose"}),
+	}
+
+	for _, c := range cases {
+		t.Run(c.desc, func(t *testing.T) {
+			if c.kind != c.table.kind {
+				t.Errorf("Table kind mismatch: %d != %d", c.kind, c.table.kind)
+			}
+			if c.name != c.table.name {
+				t.Errorf("Table name mismatch: %s != %s", c.name, c.table.name)
+			}
+			if c.version != c.table.version {
+				t.Errorf("Table version mismatch: %d != %d", c.version, c.table.version)
+			}
+			importDiff := cmp.Diff(c.imports, c.table.imports)
+			if importDiff != "" {
+				t.Errorf("Table imports mismatch:\n%s", importDiff)
+			}
+			if c.maxSID != c.table.maxSID {
+				t.Errorf("Table maxSID mismatch: %d != %d", c.maxSID, c.table.maxSID)
+			}
+			for _, badSID := range []int64{0, -10, c.maxSID + 1, c.maxSID + 128} {
+				tok, ok := c.table.BySID(badSID)
+				if ok {
+					t.Errorf("Found a token %v for a non-existent SID %d", tok, badSID)
+				}
+			}
+
+			// shuffle the tokens to randomize search a bit
+			rnd := rand.New(rand.NewSource(testSystemSymbolTableSeed))
+			rnd.Shuffle(len(c.tokens), func(i, j int) {
+				c.tokens[i], c.tokens[j] = c.tokens[j], c.tokens[i]
+			})
+			for _, expected := range c.tokens {
+				for i := 0; i < 2; i++ {
+					actualByID, ok := c.table.BySID(expected.localSID)
+					if !ok {
+						t.Error("Could not find ", expected)
```

Comment:
Not existing will show up as a specific diff making the existence check unnecessary.  I would probably kill a bit of code by reducing to:

```go
actualByID, _ := c.table.BySID(expected.localSID)
if diff := cmp.Diff(expected, actualByID, exportAll); diff != "" {
    t.Errorf("Tokens are structurally different (-expected, +found): %s", diff)
}
```
Totally a style choice.  The reduced version may be too condensed to be palatable for some.

Reply:
I get that, but this also makes sure we're returning the right bool as well.